samples/peripheral/lpuart/src/main.c

```diff
@@ -15,21 +17,44 @@ LOG_MODULE_REGISTER(app);
 #define BUF_SIZE 64
 static K_MEM_SLAB_DEFINE(uart_slab, BUF_SIZE, 3, 4);
 
-static void uart_irq_handler(const struct device *dev, void *context)
+#define TX_DATA_SIZE 5
+static const uint8_t tx_buf[TX_DATA_SIZE] = {1, 2, 3, 4, 5};
+static uint8_t rx_buf[TX_DATA_SIZE + 1];
+static uint8_t buffer_print_tx[TX_DATA_SIZE * 5 + 1];
+static uint8_t buffer_print_rx[TX_DATA_SIZE * 5 + 1];
+
+static void to_display_format(const uint8_t *src, size_t size, char *dst)
 {
-	uint8_t buf[] = {1, 2, 3, 4, 5};
+	size_t i;
+
+	for (i = 0; i < size; i++) {
+		sprintf(dst + 5 * i, "0x%02x,", src[i]);
+	}
+}
 
+static void uart_irq_handler(const struct device *dev, void *context)
+{
 	if (uart_irq_tx_ready(dev)) {
-		(void)uart_fifo_fill(dev, buf, sizeof(buf));
+		(void)uart_fifo_fill(dev, tx_buf, TX_DATA_SIZE);
 		uart_irq_tx_disable(dev);
+		LOG_INF("IRQ Tx sent %d bytes", TX_DATA_SIZE);
 	}
 
 	if (uart_irq_rx_ready(dev)) {
-		uint8_t buf[10];
-		int len = uart_fifo_read(dev, buf, sizeof(buf));
+		int len = uart_fifo_read(dev, rx_buf, sizeof(rx_buf));
 
 		if (len) {
-			printk("read %d bytes\n", len);
+			LOG_INF("IRQ Received data %d bytes", len);
+		}
+		__ASSERT(len == 1 || len == TX_DATA_SIZE, "Received unexpected data");
+		if (len == TX_DATA_SIZE) {
+			if (memcmp(tx_buf, rx_buf, len)) {
+				to_display_format(tx_buf, TX_DATA_SIZE, buffer_print_tx);
+				to_display_format(rx_buf, len, buffer_print_rx);
+				LOG_ERR("Buffer contents are different: %s", buffer_print_tx);
+				LOG_ERR("                           vs: %s", buffer_print_rx);
+				__ASSERT(false, "Buffer contents are different");
+			}
 		}
```

Comment:
Why not extract this to a helper function? The corresponding code for the async part is actually the same, only a few parameters are different. Buffers `buffer_print_tx` and `buffer_print_rx` could be local then (actually, wouldn't just one buffer be sufficient for this formatting?).

Reply:
Changed to helper function, now only single `buffer_print` is used, thanks.